src/Access/WopiProofAccessCheck.php

```diff
@@ -0,0 +1,229 @@
+   *
+   * @return array<'current'|'old', \phpseclib3\Crypt\RSA\PublicKey>
+   *   Current and old public key, or just the current if they are the same, or
+   *   empty array if none found.
+   */
+  protected function getKeys(): array {
+    // Get current and old key.
+    // Remove empty values.
+    // If both are the same, keep only the current one.
+    $public_keys = array_unique(array_filter([
+      'current' => $this->discovery->getProofKey(),
+      'old' => $this->discovery->getProofKeyOld(),
+    ]));
+    $key_objects = [];
+    foreach ($public_keys as $key_name => $key_str) {
+      $key_obj = $this->prepareKey($key_str, $key_name);
+      if ($key_obj === NULL) {
+        continue;
+      }
+      $key_objects[$key_name] = $key_obj;
+    }
+    return $key_objects;
+  }
+
+  /**
+   * Gets an RSA key object based on a string value.
+   *
+   * @param string $key_str
+   *   Key string value from discovery.xml.
+   * @param 'current'|'old' $key_name
+   *   Key name, only used for logging.
+   *
+   * @return \phpseclib3\Crypt\RSA\PublicKey|null
+   *   An RSA public key object, or NULL on failure.
+   */
+  protected function prepareKey(string $key_str, string $key_name): ?PublicKey {
+    try {
+      $key_object = PublicKeyLoader::loadPublicKey($key_str);
+    }
+    catch (\Throwable $e) {
+      $log_message = "Problem with the @name key from discovery.yml:<br>\n"
+        . Error::DEFAULT_ERROR_MESSAGE;
+      $log_args = Error::decodeException($e);
+      $log_args['@name'] = $key_name;
+      $this->logger->error($log_message, $log_args);
+      return NULL;
+    }
+    if (!$key_object instanceof PublicKey) {
+      $log_message = "Problem with the @name key from discovery.yml:<br>\n"
+        . "Expected RSA public key, found @type.";
+      $log_args = [
+        '@name' => $key_name,
+        '@type' => get_debug_type($key_object),
+      ];
+      $this->logger->error($log_message, $log_args);
+      return NULL;
+    }
+    return $key_object
+      ->withHash('sha256')
+      ->withPadding(RSA::SIGNATURE_RELAXED_PKCS1);
+  }
+
+  /**
+   * Gets the current and old signature from the request.
+   *
+   * The request will have a current and an old signature.
+   * This is to support situations when the key has been recently changed, but
+   * the cached discovery.xml still has the old key.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   Incoming request that may have a signature to be verified.
+   *
+   * @return array{current?: string, old?: string}
+   *   Current and old signature from the request, decoded and ready for use.
+   *   If they are the same, only one of them is returned.
+   *   If no signatures are found, an empty array is returned.
+   */
+  protected function getSignatures(Request $request): array {
+    // Get the current and old proof header.
+    // Remove empty values.
+    // If both are the same, keep only the current one.
```

Comment:
If the keys are not changed regularly, then we will have identical old and new key, and identical old and new signature. You will see that in the existing discovery.xml.
The only implication of keeping duplicates is that we will do redundant checks. So not a big deal, but also easy enough to avoid.

Reply:
The redundant checks would only happen in the non-happy path, so it is not really the main scenario to optimize for.
The wopi-lib has explicit vars for old and new key and proof and does the 3 checks explicitly with `&&`.
We could do that.

Currently with the `array_filter()` and then `foreach()` we silently ignore if an old key is missing in discovery.xml, or the old proof is missing in the request.
Instead, we could simply fail if the old key or proof is not present.
This would reveal if e.g. we are using the wrong xpath or wrong header name, or if the WOPI client behaves differently due to an update.